openmls/src/group/core_group/new_from_welcome.rs

```diff
@@ -11,6 +11,292 @@ use crate::{
     },
 };
 
+impl StagedCoreGroup {
+    /// Create a staged join from a welcome message. The purpose of this type is to be able to
+    /// extract information, such as the identify of who created the welcome, before joining the
+    /// group.
+    /// Note: calling this function will consume the key material for decrypting the [`Welcome`]
+    /// message, even if the caller does not turn the [`StagedCoreGroup`] into a [`CoreGroup`].
+    pub fn new_from_welcome<KeyStore: OpenMlsKeyStore>(
+        welcome: Welcome,
+        ratchet_tree: Option<RatchetTreeIn>,
+        key_package_bundle: KeyPackageBundle,
+        provider: &impl OpenMlsProvider<KeyStoreProvider = KeyStore>,
+        mut resumption_psk_store: ResumptionPskStore,
+    ) -> Result<Self, WelcomeError<KeyStore::Error>> {
+        log::debug!("CoreGroup::new_from_welcome_internal");
```

Comment:
This duplicates quite a bit of code. Can we find a way to share the code? For example calling this in `new_from_welcome` or factoring out the common code into a helper function.